pkg/provider/infisical/token.go

```diff
@@ -0,0 +1,170 @@
+	accessTokenFetchedTime   time.Time
+	accessTokenRefreshedTime time.Time
+
+	mutex sync.Mutex
+
+	accessToken  string
+	clientSecret string
+	clientID     string
+
+	apiClient api.InfisicalApis
+}
+
+func NewMachineTokenManger(apiClient api.InfisicalApis, clientID, clientSecret string) *MachineTokenManager {
+	token := &MachineTokenManager{
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		apiClient:    apiClient,
+	}
+
+	return token
+}
+
+func (t *MachineTokenManager) StartLifeCycle() error {
+	if err := t.FetchNewAccessToken(); err != nil {
+		return err
+	}
+	go t.HandleTokenLifecycle()
+	return nil
+}
+
+func (t *MachineTokenManager) HandleTokenLifecycle() {
+	// this is skip doing things in first iteration.
+	// the refresh part is not needed to trigger at first because access token is just received as on top.
+	init := false
+	for {
+		accessTokenMaxTTLExpiresInTime := t.accessTokenFetchedTime.Add(t.accessTokenMaxTTL - (5 * time.Second))
+		accessTokenRefreshedTime := t.accessTokenRefreshedTime
+
+		if accessTokenRefreshedTime.IsZero() {
+			accessTokenRefreshedTime = t.accessTokenFetchedTime
+		}
+
+		if time.Now().After(accessTokenMaxTTLExpiresInTime) {
+			Logger.Info("Infisical Authentication: machine identity access token has reached max ttl, attempting to re authenticate...")
+			err := t.FetchNewAccessToken()
+			if err != nil {
+				Logger.Error(err, "Infisical Authentication: unable to authenticate universal auth. Will retry in 30 seconds")
+
+				// wait a bit before trying again.
+				time.Sleep((30 * time.Second))
+				continue
+			}
+		} else if init {
+			err := t.RefreshAccessToken()
+			if err != nil {
+				Logger.Error(err, "Infisical Authentication: unable to refresh universal auth token because %v. Will retry in 30 seconds")
+
+				// wait a bit before trying again.
+				time.Sleep((30 * time.Second))
+				continue
+			}
+		}
+
+		if accessTokenRefreshedTime.IsZero() {
+			accessTokenRefreshedTime = t.accessTokenFetchedTime
+		} else {
+			accessTokenRefreshedTime = t.accessTokenRefreshedTime
+		}
+
+		nextAccessTokenExpiresInTime := accessTokenRefreshedTime.Add(t.accessTokenTTL - (5 * time.Second))
+		accessTokenMaxTTLExpiresInTime = t.accessTokenFetchedTime.Add(t.accessTokenMaxTTL - (5 * time.Second))
+
+		if nextAccessTokenExpiresInTime.After(accessTokenMaxTTLExpiresInTime) {
+			// case: Refreshed so close that the next refresh would occur beyond max ttl (this is because currently, token renew tries to add +access-token-ttl amount of time).
+			// example: access token ttl is 11 sec and max ttl is 30 sec. So it will start with 11 seconds, then 22 seconds but the next time you call refresh it would try to extend it to 33 but max ttl only allows 30, so the token will be valid until 30 before we need to reauth.
+			time.Sleep(t.accessTokenTTL - nextAccessTokenExpiresInTime.Sub(accessTokenMaxTTLExpiresInTime))
+		} else {
+			time.Sleep(t.accessTokenTTL - (5 * time.Second))
```

Comment:
I'm really bothered with `time.sleep` in code... did you by any chance ran a load test to see if this logic would be safe to use? I'd suggest at least testing it on a setup with 200 ExternalSecrets on 20 SecretStores with 1s refreshInterval and operator `concurrency` of 1000)

Reply:
Thread discussed above